sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/service/write/DocumentWriteService.java

```diff
@@ -0,0 +1,506 @@
+   * @param documentId Document ID.
+   * @param rows Rows of this document.
+   * @param ttl the time-to-live of the rows (seconds)
+   * @param context Execution content for profiling.
+   * @return Uni containing the {@link ResultSet} of the batch execution.
+   */
+  @WithSpan
+  public Uni<ResultSet> updateDocument(
+      String keyspace,
+      String collection,
+      String documentId,
+      List<JsonShreddedRow> rows,
+      Integer ttl,
+      ExecutionContext context) {
+    List<String> subDocumentPath = Collections.emptyList();
+    return updateDocumentInternal(
+        keyspace, collection, documentId, subDocumentPath, rows, ttl, context);
+  }
+
+  /**
+   * Updates a single document, ensuring that existing document with the same key have the given
+   * sub-path deleted.
+   *
+   * @param keyspace Keyspace to store document in.
+   * @param collection Collection the document belongs to.
+   * @param documentId Document ID.
+   * @param subDocumentPath The sub-document path to delete.
+   * @param rows Rows of this document.
+   * @param ttl the time-to-live of the rows (seconds)
+   * @param context Execution content for profiling.
+   * @return Uni containing the {@link ResultSet} of the batch execution.
+   */
+  @WithSpan
+  public Uni<ResultSet> updateDocument(
+      String keyspace,
+      String collection,
+      String documentId,
+      List<String> subDocumentPath,
+      List<JsonShreddedRow> rows,
+      Integer ttl,
+      ExecutionContext context) {
+    return updateDocumentInternal(
+        keyspace, collection, documentId, subDocumentPath, rows, ttl, context);
+  }
+
+  private Uni<ResultSet> updateDocumentInternal(
+      String keyspace,
+      String collection,
+      String documentId,
+      List<String> subDocumentPath,
+      List<JsonShreddedRow> rows,
+      Integer ttl,
+      ExecutionContext context) {
+
+    return Uni.createFrom()
+        .item(
+            () -> {
+              // check that sub path matches the rows supplied
+              // TODO this smells like a bad design. Maybe not include sub-path during shredding,
+              //  and instead add it here as a wrapper
+              checkPathMatchesRows(subDocumentPath, rows);
+
+              long timestamp = timeSource.currentTimeMicros();
+              List<QueryOuterClass.Query> queries = new ArrayList<>(rows.size() + 1);
+
+              // delete existing subpath
+              AbstractDeleteQueryBuilder deleteQueryBuilder =
+                  subDocumentPath.isEmpty()
+                      ? new DeleteDocumentQueryBuilder(documentProperties)
+                      : new DeleteSubDocumentPathQueryBuilder(
+                          subDocumentPath, false, documentProperties);
+              QueryOuterClass.Query deleteQuery =
+                  deleteQueryBuilder.buildQuery(keyspace, collection);
+              queries.add(deleteQueryBuilder.bind(deleteQuery, documentId, timestamp - 1));
```

Comment:
Unlike v1, the gRPC API doesn't have an explicit concept of a prepared statement, queries are prepared automatically.
So the separation between `buildQuery()` and `bind()` feels a bit artificial, everything could be done in a single call.
The only reason would be to avoid re-building the CQL string every time, but that means we would need to cache by keyspace+collection (+extra fields for the other builders) at the class level. I'm not convinced this would make a huge difference, feels more like premature optimization.

Any thoughts?

Reply:
Yea get it.. I would avoid caching.. We can leave it as it, or create `buildAndBind` shortcut if you like.. Although for the inser query it makes sense in creating one time and binding different values.. I guess this is a nice option..